src/emqx_protocol.erl

```diff
@@ -277,36 +285,30 @@ process_packet(?CONNECT_PACKET(
                                          keepalive    = Keepalive,
                                          conn_props   = ConnProps,
                                          will_topic   = WillTopic,
-                                         will_msg     = WillMsg,
+                                         will_msg     = emqx_packet:will_msg(Connect),
                                          is_bridge    = IsBridge,
                                          connected_at = os:timestamp()}),
-    connack(
-      case check_connect(Connect, PState1) of
-          {ok, PState2} ->
-              case authenticate(credentials(PState2), Password) of
-                  {ok, IsSuper} ->
-                      %% Maybe assign a clientId
-                      PState3 = maybe_assign_client_id(PState2#pstate{is_super = IsSuper}),
-                      %% Open session
-                      case try_open_session(PState3) of
-                          {ok, SPid, SP} ->
-                              PState4 = PState3#pstate{session = SPid, connected = true},
-                              ok = emqx_cm:register_connection(client_id(PState4), attrs(PState4)),
-                              %% Start keepalive
-                              start_keepalive(Keepalive, PState4),
-                              %% Success
-                              {?RC_SUCCESS, SP, PState4};
-                          {error, Error} ->
-                              ?LOG(error, "Failed to open session: ~p", [Error], PState1),
-                              {?RC_UNSPECIFIED_ERROR, PState1}
-                    end;
-                  {error, Reason} ->
-                      ?LOG(error, "Username '~s' login failed for ~p", [Username, Reason], PState2),
-                      {?RC_NOT_AUTHORIZED, PState1}
-              end;
-          {error, ReasonCode} ->
-              {ReasonCode, PState1}
-      end);
+
+    case check_connect(Connect, PState1) of
+        {ok, PState2} ->
+            EnhancedAuth = is_map(ConnProps) andalso maps:is_key('Authentication-Method', ConnProps),
+            if 
+                EnhancedAuth ->
+                    process_enhanced_auth(maps:get('Authentication-Method', ConnProps, undefined), 
+                                          maps:get('Authentication-Data', ConnProps, undefined),
+                                          PState2);
+                true ->
+                    case authenticate(credentials(PState2), Password) of
+                        {ok, IsSuper} ->
+                            connack(open_session(PState2#pstate{is_super = IsSuper, auth_state = connected}));
+                        {error, Reason} ->
+                            ?LOG(error, "Username '~s' login failed for ~p", [Username, Reason], PState2),
+                            connack({?RC_NOT_AUTHORIZED, PState2})
+                    end
+            end;
+        {error, ReasonCode} ->
+            connack({ReasonCode, PState1})
```

Comment:
`connack/1` should now be changed to `connack/2`
i.e. use two args instead of a 2-elements tuple as a single arg.

Reply:
never mind, just notice line 303 is still the old style call.